src/tracer.zig

```diff
@@ -0,0 +1,332 @@
+    main,
+    tracer,
+    tree: struct {
+        tree_name: []const u8,
+    },
+};
+
+const SpanId = u64;
+
+pub const SpanStart = struct {
+    id: SpanId,
+    start_time_ns: u64,
+    group: EventGroup,
+    event: Event,
+};
+
+const Span = struct {
+    id: SpanId,
+    start_time_ns: u64,
+    end_time_ns: u64,
+    group: EventGroup,
+    event: Event,
+};
+
+const span_count_max = 1 << 20;
+
+const log_path = "./tracer.json";
+
+pub fn init(allocator: Allocator) !void {
+    if (backend == .none) return;
+    assert(!is_initialized);
+    spans = try std.ArrayList(Span).initCapacity(allocator, span_count_max);
+
+    switch (backend) {
+        .none => unreachable,
+        .perfetto => {
+            const file = try std.fs.cwd().createFile(log_path, .{ .truncate = true });
+            defer file.close();
+
+            try file.writeAll(
+                \\{"traceEvents":[
+                \\
+            );
+        },
+    }
+
+    is_initialized = true;
+}
+
+pub fn deinit(allocator: Allocator) void {
+    _ = allocator;
+
+    if (backend == .none) return;
+    assert(is_initialized);
+
+    flush();
+    assert(flush_slot == null);
+    spans.deinit();
+    is_initialized = false;
+}
+
+pub fn start(slot: *?SpanStart, event_group: EventGroup, event: Event) void {
+    if (backend == .none) return;
+    assert(is_initialized);
+
+    // The event must not have already been started.
+    assert(slot.* == null);
+
+    slot.* = .{
+        .id = span_id_next,
+        .start_time_ns = timer.monotonic(),
+        .group = event_group,
+        .event = event,
+    };
+    span_id_next += 1;
+}
+
+pub fn end(slot: *?SpanStart, event_group: EventGroup, event: Event) void {
+    if (backend == .none) return;
+    assert(is_initialized);
+
+    // The event must have already been started.
+    const span_start = &slot.*.?;
+    assert(std.meta.eql(span_start.group, event_group));
+    assert(std.meta.eql(span_start.event, event));
+
+    // Make sure we have room in spans.
+    if (spans.items.len >= span_count_max) flush();
+    assert(spans.items.len < span_count_max);
+
+    spans.appendAssumeCapacity(.{
+        .id = span_start.id,
+        .start_time_ns = span_start.start_time_ns,
+        .end_time_ns = timer.monotonic(),
+        .group = span_start.group,
+        .event = span_start.event,
+    });
+    slot.* = null;
+}
+
+pub fn flush() void {
+    if (backend == .none) return;
+    assert(is_initialized);
+
+    if (spans.items.len == 0) return;
+
+    start(&flush_slot, .tracer, .tracer_flush);
+    flush_or_err() catch |err| {
+        log.err("Could not flush tracer log, discarding instead: {}", .{err});
+    };
+    spans.shrinkRetainingCapacity(0);
+    end(&flush_slot, .tracer, .tracer_flush);
+}
+
+fn flush_or_err() !void {
+    switch (backend) {
+        .none => unreachable,
+        .perfetto => {
+            const file = try std.fs.cwd().openFile(log_path, .{ .read = false, .write = true });
```

Comment:
Rather than re-opening the file every flush, maybe it could be opened exactly once in `init()` and closed in `deinit()`? Also we could wrap the file in an `io.BufferedWriter`.